### ISSUE TITLE
ARM Mac 16-core Neural Engine

### ISSUE THREAD

**Issue by fire** [pytorch/pytorch#47688]
## 🚀 Feature
Support 16-core Neural Engine in PyTorch

## Motivation

PyTorch should be able use Apple 16-core Neural Engine as the backing system.

## Pitch

Since the ARM macs have uncertain support for external GPUS. It would be wise to support the Apple Neural Engine.

## Alternatives

Use AMD64 Macs.

## Additional context

https://www.apple.com/ca/mac-mini/

**Comment by malfet** [comment]
Thank you for your suggestion. If you already have a PR in mind that partially implements NeuralEngine integration, please do not hesitate to send PR for review.

**Comment by lqf96** [comment]
@malfet The very first problem I can think of is how do we want to accelerate PyTorch on macOS computers. From my perspective there are three possible ways:

1. Write Metal compute shaders to accelerate PyTorch tensor operators. Works like #46112 have already gone down this approach but the drawback is that we can only make use of the GPU but not the neural engine.
2. Create a standalone CoreML backend for PyTorch. This may not be appropriate because in PyTorch tensor operations are performed eagerly unless JIT is used, while in CoreML you must first create the computation graph.
3. Implement XLA on top of CoreML, since both are based on computation graphs, and then use [PyTorch XLA](https://github.com/pytorch/xla) to provide support for macOS computers.

**Comment by rshpeley** [comment]
Current M1 chips have a unified memory limit of 16GB max (with 8GB more likely) with no option to expand. The GPU is somewhat better in performance than a GTX 1050 Ti and the neural engine cores could bring up the compute capacity to about the level of a 1070 at a guess. 

With 8GB system ram, free memory likely won't be more than 5-6GB. So it's performance will be decent, and a step up in laptops, but limited. That said, `torch.distributed`, together with PyTorch/XLA (modded as per @lqf96), would fill mid level requirements and could be more interesting.

Maybe petition Apple to do the PR? They know CoreML better than anyone. And a cluster of 100 mac minis would be so cool, eh?

**Comment by leovinus2001** [comment]
> @malfet The very first problem I can think of is how do we want to accelerate PyTorch on macOS computers. From my perspective there are three possible ways:

 I also love to have a proper macOS PyTorch using Metal/BNNS/Accelerate but specifically Metal to have TFLOPS training and inference performance on my iMac 64GB with Radeon Pro 580 8 GB.

> 1. Write Metal compute shaders to accelerate PyTorch tensor operators. Works like #46112 have already gone down this approach but the drawback is that we can only make use of the GPU but not the neural engine.

My impression (which might be a few months outdated, sorry) is that PyTorch < 1.7 only has some Metal  support in the iOS inference backend as to be not "too slow on CNNs" compared to CoreML on iOS and same device. What would be ideal is a macOS Metal/MPS support for training and inference throughout ATen, auto differentiate, all PyTorch tensors using Metal blocks like LSTM/CNN. That should give a training just as fast as with NVIDIA and abstracts the M1/AMD GPU/Neural details into Metal/MPS drivers. For inference, a plain conversion to CoreML on iOS and macOS should give the fastest inference. The converter coremltoolsv4 has come a long way already.

> 2. Create a standalone CoreML backend for PyTorch. This may not be appropriate because in PyTorch tensor operations are performed eagerly unless JIT is used, while in CoreML you must first create the computation graph.

Again, the coremltools v4 have gone a long way in the Pytorch to CoreML model conversion for inference. PyTorch main limitation is that >training< on the Mac never uses the GPU via Metal/MPS and even used to ignore Accelerate in favor of Intel libraries (was it 1.2 or 1.3 ? ) even though Apple can optimize more for CPU performance (GEMM) on M1 and Intel.

Years ago, I wrote an unpublished proof-of-concept Metal/MPS replacement for a "look alike" Linear layer to measure speed improvements in inference for Accelerate vs Metal on macOS. Fun, but too much effort to make a PR for PyTorch at the time.

I'd love to see a proper Mac GPU training with Metal on either Intel or M1 but am unsure how to make it happen. Effort for a small team, right?

**Comment by senderle** [comment]
Does anyone know how apple's custom tensor flow build supports the neural engine / GPU? My hunch is that they must have implemented a custom backend for XLA supporting the M1 as a target architecture. In which case building M1 support into PyTorch _should_ doable in the same way TPU support was.

Only catch is that apple would have to release the source, which AFAICT hasn't happened yet. Or else somebody would need to replicate the work on XLA.

Maybe I'm wrong about it being this simple? But apple managed solid M1 support so quickly — it seems likely that this wasn't a major overhaul, but a pretty modular change.

**Comment by SharanSMenon** [comment]
Will they add support for the Neural Engine?

**Comment by SharanSMenon** [comment]
Maybe you could just use MLCompute. I don't know too much.

**Comment by migueldeicaza** [comment]
Two interesting observations:

(a) The folks at OctoML wrote Metal-based kernels that outperform Apple's CoreML: https://venturebeat.com/2020/12/16/octoml-optimizes-apache-tvm-for-apples-m1-beats-core-ml-4-by-29/

(b) Given that there is already a Metal backend for iOS, that this Metal backend could be repurposed for training on Mac with their GPUs.

**Comment by JacopoMangiavacchi** [comment]
Hey @migueldeicaza it's my understanding both (a) and (b) only support inferencing, not training, but apple officially announced a Metal TF PluggableDevice so apparently this is the way to go, other than mlcompute.

**Comment by alelordelo** [comment]
PyTorch on Mac M1 would be awesome!

Any updated on that?

**Comment by PeterL1n** [comment]
1. I think writing custom Metal kernels are a great approach. We have full controllability in everything. We can implement both forward-pass and backward-pass for training and inference if needed. If it only supports GPU then let it be. It is Apple's problem not having a neural engine programmable and hopefully, that will change in the future. If Apple provides standard libraries we can invoke theirs like CuDNN.

2. CoreML is very limited in functionality. For example, its interpolate/resize does not support dynamic resolutions, etc. It is already a pain when exporting the PyTorch model to CoreML since it has so many ops not supported. I think building on CoreML will be very limited.

I am just curious since we already have an ongoing effort to write custom Metal for iOS. What makes it so hard to enable it on MacOS? Any plan to enable that and ship MacOS binary with Metal support? It would be great even if the current support is limited and only for inference.

**Comment by jucor** [comment]
With the new M1Max announced today, I must say I'm quite drooling for Pytorch support :)

**Comment by almson** [comment]
The thought that's piqued my interest is training on an M1 Max with 64GB of memory, for those times when I want to try an idea that won't won't fit into my GPU. 

Ironically, a 64GB MacBook is currently the same price[^1] as a 24GB Nvidia GPU and not too far in TFLOPS (and far more energy efficient). One can even imagine attaching Infiniband or [ExpressFabric](https://www.broadcom.com/products/pcie-switches-bridges/expressfabric) over the Thunderbolt port.

Then again I plan to run Linux on my M1 Mac, so I'd actually be interested in Vulkan or possibly lower-level support. The Asahi Linux effort is doing extremely well, and I urge everyone to help sponsor them. They're doing a lot of reverse-engineering, including of the GPU. Perhaps it would be more productive to just write a CUDA (or HIP?) compatibility layer which might also run on macOS.

[^1]: Fantasizing here, but if one were planning to build an M1 cluster, then selling the unneeded parts (screen, keyboard, battery, ports) could perhaps half its cost.

**Comment by ltto** [comment]
都2022年了 还不支持吗

**Comment by GeekOfComputer01100101** [comment]
The neural engine only supports 16-bit half precision and 8-bit integers. That's too few exponent bits to be differentiable. Even if the APIs did support it, it would be physically impossible to use the NPU for training. Sorry to disappoint, but it's a fundamental limitation with the microarchitecture.

**Comment by rshpeley** [comment]
multiprecision, and mixed precision, training can use fp16 of the neural processor and fp32 of the cpu. 

apple’s neural processor supports multi precision but not likely mixed precision without getting info on programming the inners the anp.

https://towardsdatascience.com/understanding-mixed-precision-training-4b246679c7c4

for one reference on mixed precision.

so it looks entirely plausible to use the np on the m1 for training — as apple does.

but getting that to work in a cluster of m1s? probably only when apple supports it directly.

— ray

**Comment by GeekOfComputer01100101** [comment]
There are two ways to access ANE - MLCompute and CoreML. MLCompute only allows using it for inference, and there's no way to access the raw ML primitives such as gradient of depthwise convolution. When using CoreML, I have no idea how severe the latency of making an `MLModel` is. It definitely wouldn't be feasible in eager mode, but maybe possible in graph mode?

**Comment by JudeDavis1** [comment]
ANE updates on this issue? (Sorry, couldn't resist the pun) Do we have support for the neural engine?

**Comment by m-grig** [comment]
Would love to see this come to fruition. One of the major drawbacks of the Apple Silicon is the machine learning limitations, when it's such a strong option for every other developer task.

**Comment by amztc34283** [comment]
Oh so I have not been running the neural engine at all?!?!?!?! I have been tricked by Apple's marketing :(
